README.md
# Angular Form Validator

[![GitHub](https://img.shields.io/github/license/truonghungit/ngx-validator)](./LICENSE)
[![commitizen](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)]()
[![PRs](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)]()
[![release](https://github.com/truonghungit/ngx-validator/actions/workflows/release.yml/badge.svg)](https://github.com/truonghungit/ngx-validator/actions/workflows/release.yml)
[![npm](https://img.shields.io/npm/v/@popeyelab/ngx-validator)](https://www.npmjs.com/package/@popeyelab/ngx-validator)
[![npm](https://img.shields.io/npm/dw/@popeyelab/ngx-validator)]()

Get validation messages with zero coding in less than 30 seconds (seriously)

Created by Angular developer for Angular developers with ❤️.

## Table of Contents

- [Why](#why)
- [Demo](https://ngx-validator-showcase.vercel.app/)
- [Quick Start](#quick-start)
- [Validation Messages](#validation-messages)
- [When Does Validation Run](#when-does-validation-run)
- [Validators](#validators)
- [Changelog](#changelog)
- [License](#license)

## Why

### Without **@popeyelab/ngx-validator**

To declarative validation error messages for reactive forms. Typically you'd do something like this:

```html
<form [formGroup]="form" #ngForm="ngForm">
  <div>
    <label>Your Name</label>
    <input type="text" formControlName="name" />

    <div *ngIf="form.controls['name'].invalid && ((form.controls['name'].dirty && form.controls['name'].touched) || ngForm.submitted)">
      <div *ngIf="form.controls['name'].hasError('required')">Name is required</div>
    </div>
  </div>

  <div>
    <label>Email Address</label>
    <input type="email" formControlName="email" />

    <div *ngIf="form.controls['email'].invalid && ((form.controls['email'].dirty && form.controls['email'].touched) || ngForm.submitted)">
      <div *ngIf="form.controls['email'].hasError('required')">Email is required</div>
      <div *ngIf="form.controls['email'].hasError('email')">Email is invalid</div>
    </div>
  </div>

  <div>
    <label>Password</label>
    <input type="password" formControlName="password" />
    <div *ngIf="form.controls['password'].invalid && ((form.controls['password'].dirty && form.controls['password'].touched) || ngForm.submitted)">
      <div *ngIf="form.controls['password'].hasError('required')">Password is required</div>
      <div *ngIf="form.controls['password'].hasError('minlength')">Password should have at least 6 characters</div>
    </div>
  </div>

  <div>
    <label>Confirm Password</label>
    <input type="password" formControlName="confirmPassword" />
    <div *ngIf="form.controls['confirmPassword'].invalid && ((form.controls['confirmPassword'].dirty && form.controls['confirmPassword'].touched ) || ngForm.submitted)">
      <div *ngIf="form.controls['confirmPassword'].hasError('required')">Confirm password is required</div>
      <div *ngIf="form.controls['confirmPassword'].hasError('equalTo')">Confirm password not match</div>
    </div>
  </div>

  <div>
    <button type="submit">Submit</button>
  </div>
</form>
```

### With **@popeyelab/ngx-validator**, we don't need to do any things, it just simple like this

```html
<form [formGroup]="form">
  <div>
    <label>Your Name</label>
    <input type="text" formControlName="name" />
  </div>

  <div>
    <label>Email Address</label>
    <input type="email" formControlName="email" />
  </div>

  <div>
    <label>Password</label>
    <input type="password" formControlName="password" />
  </div>

  <div>
    <label>Confirm Password</label>
    <input type="password" formControlName="confirmPassword" />
  </div>

  <div>
    <button type="submit">Submit</button>
  </div>
</form>
```

[Back to top](#table-of-contents)

## Quick Start

Follow these steps to get started with **Ngx Validator**.

> Actually we only install package then add the `FormValidatorModule` to your module

### 1. Install **@popeyelab/ngx-validator** package:

Installing with `npm`

```
npm install @popeyelab/ngx-validator --save
```

Installing with `yarn`

```
yarn add @popeyelab/ngx-validator
```

### 2. Add `FormValidatorModule` module to your module

_app.module.ts_

```typescript
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { ReactiveFormsModule } from '@angular/forms';

import { FormValidatorModule } from '@popeyelab/ngx-validator';

import { AppComponent } from './app.component';

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule, BrowserAnimationsModule, ReactiveFormsModule, FormValidatorModule, AppRoutingModule],
  bootstrap: [AppComponent],
})
export class AppModule {}
```

### 3. Define the form and validators

_app.component.ts_

```typescript
import { Component } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { equalTo } from '@popeyelab/ngx-validator';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
})
export class AppComponent {
  form = new FormGroup({
    name: new FormControl('', [Validators.required]),
    email: new FormControl('', [Validators.required, Validators.email]),
    password: new FormControl('', [Validators.required, Validators.minLength(6)]),
    confirmPassword: new FormControl('', [Validators.required, equalTo('password')]),
  });
}
```

_app.component.html_

```html
<form [formGroup]="form">
  <div>
    <label>Your Name</label>
    <input type="text" formControlName="name" />
  </div>

  <div>
    <label>Email Address</label>
    <input type="email" formControlName="email" />
  </div>

  <div>
    <label>Password</label>
    <input type="password" formControlName="password" />
  </div>

  <div>
    <label>Confirm Password</label>
    <input type="password" formControlName="confirmPassword" />
  </div>

  <div>
    <button type="submit">Submit</button>
  </div>
</form>
```

> As you can see, we don't have to do anymore in html template. The `Ngx Validator` will know how to display the validation messages to the user interface.

[Back to top](#table-of-contents)

## Validation Messages

The library will capture the error (`error: ValidationErrors`) from the from control, then show the validation messages to UI based on the default validation messages as bellow.

| Error         | Message                                               |
| ------------- | ----------------------------------------------------- |
| required      | This field is required                                |
| email         | Email is invalid                                      |
| max           | Value should be less than or equal to {{ max }}.      |
| min           | Value should be greater than or equal to {{ min }}.   |
| maxlength     | {{ requiredLength }} characters are allowed.          |
| minlength     | Should have at least {{ requiredLength }} characters. |
| pattern       | Invalid pattern. Please review your input.            |
| range         | Value should be in the range of {{ range }}           |
| url           | URL is invalid.                                       |
| unknown error | [This field is invalid]                               |

### Custom validation Messages

We have the following options for changing the validation message:

#### Option 1: Use form validation configuration

```typescript
import { FormValidatorModule, FormValidatorConfig } from '@popeyelab/ngx-validator';

const formValidatorConfig: Partial<FormValidatorConfig> = {
  validationMessages: {
    required: 'This field is required.',
    email: 'Email is invalid',
    max: 'Value should be less than or equal to {{ max }}.',
    maxlength: '{{ requiredLength }} characters are allowed.',
    min: 'Value should be greater than or equal to {{ min }}.',
    minlength: 'Should have at least {{ requiredLength }} characters.',
    pattern: 'Invalid pattern. Please review your input.',
    range: 'Value should be in the range of {{ range }}',
    url: 'URL is invalid.',
  },
};

@NgModule({
  imports: [
    // ...
    FormValidatorModule.configure(formValidatorConfig),
  ],
})
export class AppModule {}
```

#### Option 2: Use `validationMessages` input of `FormControlValidatorDirective`

_app.component.html_

```html
<form [formGroup]="form">
  <div>
    <label>Your Name</label>
    <input type="text" formControlName="name" [validationMessages]="{required: 'Please enter your name'}" />
  </div>

  <div>
    <label>Email Address</label>
    <input type="email" formControlName="email" [validationMessages]="{required: 'Please enter your email', email: 'Please enter valid email'}" />
  </div>
</form>
```

#### Option 3: Pass validation messages to validator function

Default Angular build-in validator functions does not support passing validation messages, so you should use the wrapped validator functions from `@popeyelab/ngx-validator`;

_app.component.ts_

```typescript
import { Component } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { required, equalTo, email } from '@popeyelab/ngx-validator';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
})
export class AppComponent {
  form = new FormGroup({
    name: new FormControl('', [required('Please enter your name')]),
    email: new FormControl('', [required('Please enter your email'), email('Email is invalid')]),
    password: new FormControl('', [Validators.required, Validators.minLength(6)]),
    confirmPassword: new FormControl('', [Validators.required, equalTo('password', 'Confirm password does not match')]),
  });
}
```

### Custom Validation Messages for Your Custom Validator

If you have a custom validator, that returns an error like the example bellow

```typescript
function color(color: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return control.value?.toLowerCase() === color ? null : { wrongColor: { requiredColor: color, actual: control.value } };
  };
}

//...
const control = new FormControl('green', color('blue'));
```

You can also add the validation message for custom form validator by

#### Option 1: Use form validation configuration

```typescript
import { FormValidatorConfig } from '@popeyelab/ngx-validator';

const formValidatorConfig: Partial<FormValidatorConfig> = {
  validationMessages: {
    // ...
    wrongColor: 'Invalid color',
  },
};
```

#### Option 2: Use `validationMessages` input of `FormControlValidatorDirective`

_app.component.html_

```html
<form [formGroup]="form">
  <div>
    <label>Your Color</label>
    <input type="text" formControlName="color" [validationMessages]="{wrongColor: 'Invalid color'}" />
  </div>
</form>
```

#### Option 3: Return error message in your custom validator function

_app.component.html_

```typescript
function color(color: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return control.value?.toLowerCase() === color
      ? null
      : {
          wrongColor: {
            requiredColor: color,
            actual: control.value,
            message: 'Invalid color',
          },
        };
  };
}
```

[Back to top](#table-of-contents)

## When Does Validation Run

By default, ngx-validator will run validation method when:

#### Form control dirty & touched

> A controls is `dirty` if users have changed the value of the control on UI.
>
> A controls is `touched` if users have focused on the control and then focused on something else.
> For example by clicking into the control and then pressing tab or clicking on another control in the form.

#### OR Whenever users attempt to submit a form.

You can control when ngx-validator runs validation by changing the configuration depending on your needs.

```typescript
const formValidatorConfig: Partial<FormValidatorConfig> = {
  validateOn: (status) => {
    return status.touched || status.submited;
  },
};

@NgModule({
  imports: [FormValidatorModule.configure(formValidatorConfig)],
})
export class AppModule {}
```

## Validators

### Angular built-in validators

Wrapped angular built-in validators, that help we can pass the custom validation message to the validator function when creating the form control

See more details in the [Angular docs](https://angular.io/api/forms/Validators)

#### min

Validator that requires the control's value to be greater than or equal to the provided number.

`min(min: number, message: string): ValidatorFn`

##### Parameters

| Name    | Type   |
| ------- | ------ |
| min     | number |
| message | string |

##### Usage:

```typescript
import { min } from '@popeyelab/ngx-validator';

// ...
const control = new FormControl(2, [min(10, 'Value should be greater than or equal to 10')]);
console.log(control.errors); // {min: {min: 10, actual: 2, message: 'Value should be greater than or equal to 10'}}
```

#### max

Validator that requires the control's value to be less than or equal to the provided number.

`max(max: number, message: string): ValidatorFn`

##### Parameters

| Name    | Type   |
| ------- | ------ |
| max     | number |
| message | string |

##### Usage:

```typescript
import { max } from '@popeyelab/ngx-validator';

// ...
const control = new FormControl(20, max(10, 'Value should be less than or equal to 10'));
console.log(control.errors); // {max: {max: 10, actual: 20, message: 'Value should be less than or equal to 10'}}
```

#### required

Validator that requires the control have a non-empty value.

`required(message: string): ValidatorFn`

##### Parameters

| Name    | Type   |
| ------- | ------ |
| message | string |

##### Usage:

```typescript
import { required } from '@popeyelab/ngx-validator';

// ...
const control = new FormControl('', required('This field is required'));
console.log(control.errors); // {required: {message: 'This field is required'}}
```

#### requiredTrue

Validator that requires the control's value be true. This validator is commonly used for required checkboxes.
`requiredTrue(message: string): ValidatorFn`

##### Parameters

| Name    | Type   |
| ------- | ------ |
| message | string |

##### Usage:

```typescript
import { requiredTrue } from '@popeyelab/ngx-validator';

// ...
const control = new FormControl(false, requiredTrue('This field is required'));
console.log(control.errors); // {required: {message: 'This field is required'}}
```

#### email

Validator that requires the control's value pass an email validation test.
`email(message: string): ValidatorFn`

##### Parameters

| Name    | Type   |
| ------- | ------ |
| message | string |

##### Usage:

```typescript
import { email } from '@popeyelab/ngx-validator';

// ...
const control = new FormControl('example email', email('This email is invalid'));
console.log(control.errors); // {email: {message: 'This email is invalid'}}
```

#### minLength

Validator that requires the length of the control's value to be greater than or equal to the provided minimum length.
`minLength(minLength: number, message: string): ValidatorFn`

##### Parameters

| Name      | Type   |
| --------- | ------ |
| minLength | number |
| message   | string |

##### Usage:

```typescript
import { minLength } from '@popeyelab/ngx-validator';
// ...

const control = new FormControl('hello', minLength(10, 'Should have at least 10 characters'));
console.log(control.errors); // {minlength: {requiredLength: 10, actualLength: 5, message: 'Should have at least 10 characters'}}
```

#### maxLength

Validator that requires the length of the control's value to be less than or equal to the provided maximum length.
`maxLength(maxLength: number, message: string): ValidatorFn`

##### Parameters

| Name      | Type   |
| --------- | ------ |
| maxLength | number |
| message   | string |

##### Usage:

```typescript
import { maxLength } from '@popeyelab/ngx-validator';
// ...

const control = new FormControl('hello world', maxLength(10, 'The max length of 10 characters is reached'));
console.log(control.errors); // {maxlength: {requiredLength: 10, actualLength: 11, message: 'The max length of 10 characters is reached'}}
```

#### pattern

Validator that requires the control's value to match a regex pattern.
`pattern(pattern: string | RegExp, message: string): ValidatorFn`

##### Parameters

| Name    | Type   |
| ------- | ------ | ------ |
| pattern | string | RegExp |
| message | string |

##### Usage:

```typescript
import { pattern } from '@popeyelab/ngx-validator';
// ...

const onlyNumber = /^[0-9]+$/;
const control = new FormControl('Hello', pattern(onlyNumber, 'Please input numeric characters only'));

console.log(control.errors); // {pattern: {requiredPattern: '/^[0-9]+$/', actualValue: 'Hello', message: 'Please input numeric characters only'}}
```

### Additional Validators

Provides a set of common validators that can be used by form controls.

#### range

Validator that requires the control's value to be in the provided range

`range(range: [number, number], message: string): ValidatorFn`

##### Parameters

| Name    | Type             |
| ------- | ---------------- |
| range   | [number, number] |
| message | string           |

##### Usage:

```typescript
import { range } from '@popeyelab/ngx-validator';

// ...
const [min, max] = [0, 10];
const control = new FormControl(11, range([min, max], 'Value should be in the range of 0 to 10'));
console.log(control.errors); // {range: {range: [0, 10], actual: 11,  message: 'Value should be in the range of 0 to 10'}}
```

#### url

Validator that requires the control's value pass an URL.
`url(message?: string): ValidatorFn`

##### Parameters

| Name    | Type   |
| ------- | ------ |
| message | string |

##### Usage:

```typescript
import { url } from '@popeyelab/ngx-validator';

// ...
const control = new FormControl('example url', url('Invalid URL'));
console.log(control.errors); // {url: {message: 'Invalid URL'}}
```

### equal

Validator that requires the control's value equal to the provided value.
`equal(value: string, message?: string): ValidatorFn`

##### Parameters

| Name    | Type   |
| ------- | ------ |
| value   | string |
| message | string |

##### Usage:

```typescript
import { equal } from '@popeyelab/ngx-validator';

// ...
const control = new FormControl('hello', equal('hello world', 'Value should be equal to hello world'));
```

### equalTo

Validator that requires the control's value equal to the value of provided control.
`equalTo(controlNamePath: string, message?: string): ValidatorFn`

##### Parameters

| Name            | Type   |
| --------------- | ------ |
| controlNamePath | string |
| message         | string |

##### Usage:

```typescript
import { Component } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { equalTo } from '@popeyelab/ngx-validator';

// ...

form = new FormGroup({
  password: new FormControl(''),
  confirmPassword: new FormControl('', [equalTo('password', 'Confirm Password does not match')]),
});
```


### noWhitespace

The validator requires that the value of the control not be empty or contain only spaces
`noWhitespace(message?: string): ValidatorFn`

##### Parameters

| Name            | Type   |

CHANGELOG.md
## [2.6.1](https://github.com/truonghungit/ngx-validator/compare/v2.6.0...v2.6.1) (2023-11-28)

# [2.6.0](https://github.com/truonghungit/ngx-validator/compare/v2.5.5...v2.6.0) (2023-11-28)


### Features

* **validator:** add new noWhitespace validator function ([ccf9840](https://github.com/truonghungit/ngx-validator/commit/ccf984021d244f1f61f2f72b83ca5d4a1459232f))

## [2.5.5](https://github.com/truonghungit/ngx-validator/compare/v2.5.4...v2.5.5) (2023-05-25)


### Bug Fixes

* :bug: resolve version conflict issue ([29f4021](https://github.com/truonghungit/ngx-validator/commit/29f40210c52ae30a05f3993034c56787d4d598f1))

## [2.5.4](https://github.com/truonghungit/ngx-validator/compare/v2.5.3...v2.5.4) (2023-05-23)

## [2.5.3](https://github.com/truonghungit/ngx-validator/compare/v2.5.2...v2.5.3) (2023-05-23)

## [2.5.2](https://github.com/truonghungit/ngx-validator/compare/v2.5.1...v2.5.2) (2023-05-22)

## [2.5.1](https://github.com/truonghungit/ngx-validator/compare/v2.5.0...v2.5.1) (2023-05-22)

# [2.5.0](https://github.com/truonghungit/ngx-validator/compare/v2.4.1...v2.5.0) (2023-05-20)


### Features

* :sparkles: add equal and equalTo validator functions ([3bb8f77](https://github.com/truonghungit/ngx-validator/commit/3bb8f776d9528f8e3fce7df5e630d47be04d34a7))

## [2.4.1](https://github.com/truonghungit/ngx-validator/compare/v2.4.0...v2.4.1) (2023-05-19)


### Bug Fixes

* :bug: add default validationMessagesComponent ([6c18d8c](https://github.com/truonghungit/ngx-validator/commit/6c18d8c39166212d0e69db140b9597bfa55b8bc6))

# [2.4.0](https://github.com/truonghungit/ngx-validator/compare/v2.3.0...v2.4.0) (2023-05-19)


### Features

* :sparkles: add default validation message for range and url ([b7e9522](https://github.com/truonghungit/ngx-validator/commit/b7e95221223f39608db27b1a4b7e5f88fae4326c))

# [2.3.0](https://github.com/truonghungit/ngx-validator/compare/v2.2.0...v2.3.0) (2023-05-19)


### Bug Fixes

* :bug: resolve display duplicate validation message ([728b7d1](https://github.com/truonghungit/ngx-validator/commit/728b7d14dccebe88d2610564746380a9040279d9))


### Features

* :sparkles: add transition message animation, remove angular material import ([05a241f](https://github.com/truonghungit/ngx-validator/commit/05a241f4d08d053f7d2dd5f8592d66ed05772b53))
* :sparkles: make the message parameter in the url validator function is optional ([1cebbc6](https://github.com/truonghungit/ngx-validator/commit/1cebbc6f3896153bd6e0eb6cbb59ff4717e97048))

# [2.2.0](https://github.com/truonghungit/ngx-validator/compare/v2.1.0...v2.2.0) (2023-05-18)


### Features

* :sparkles: add url validator function ([c8ee3c8](https://github.com/truonghungit/ngx-validator/commit/c8ee3c8702c989d7434801ffe8bafc1f42f4751f))

# [2.1.0](https://github.com/truonghungit/ngx-validator/compare/v2.0.0...v2.1.0) (2023-05-18)


### Features

* add validate for angular material UI ([b5347ee](https://github.com/truonghungit/ngx-validator/commit/b5347ee4189196e95f2558d6cbbb6ee0417a8633))

# [2.0.0](https://github.com/truonghungit/ngx-validator/compare/v1.0.0...v2.0.0) (2023-05-16)


### Features

* :sparkles: update validate function to suport bootstrap ([db22ec7](https://github.com/truonghungit/ngx-validator/commit/db22ec762582f0d2c6eefcb92cd08f79465341ba))


### BREAKING CHANGES

* add bootstrap validation message, change validator target selector

# 1.0.0 (2023-05-15)


### chore

* 🤖 create initial files ([7902d4a](https://github.com/truonghungit/ngx-validator/commit/7902d4ae22b1a5498f51236b8c18623a774ed388))


### BREAKING CHANGES

* remove NgxValidator class and use functions instead


projects/ngx-validator/src/lib/validators/validators.ts
import {
  Validators as NgValidators,
  ValidatorFn,
  AbstractControl,
  ValidationErrors,
} from '@angular/forms';

const URL_REGEXP = /^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})).?)(?::\d{2,5})?(?:[/?#]\S*)?$/i;

function isObject(arg: any): boolean {
  return arg !== null && typeof arg === 'object' && !Array.isArray(arg);
}

function validate(
  validatorFn: ValidatorFn,
  validatorKey: string,
  control: AbstractControl,
  message: string
): ValidationErrors | null {
  const validationErrors = validatorFn(control);

  if (message) {
    if (validationErrors?.[validatorKey]) {
      if (!isObject(validationErrors[validatorKey])) {
        validationErrors[validatorKey] = {};
      }

      validationErrors[validatorKey].message = message;
    }
  }

  return validationErrors;
}

function isEmptyInputValue(value: any): boolean {
  /**
   * Check if the object is a string or array before evaluating the length attribute.
   * This avoids falsely rejecting objects that contain a custom length attribute.
   * For example, the object {id: 1, length: 0, width: 0} should not be returned as empty.
   */
  return (
    value == null ||
    ((typeof value === 'string' || Array.isArray(value)) && value.length === 0)
  );
}

/**
 * @description
 * Validator that requires the control's value to be greater than or equal to the provided number.
 *
 * @usageNotes
 *
 * ### Validate against a minimum of 3
 *
 * ```typescript
 * const control = new FormControl(2, min(3));
 *
 * console.log(control.errors); // {min: {min: 3, actual: 2}}
 * ```
 *
 * @returns A validator function that returns an error map with the
 * `min` property if the validation check fails, otherwise `null`.
 *
 */
export function min(min: number, message: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return validate(NgValidators.min(min), 'min', control, message);
  };
}

/**
 * @description
 * Validator that requires the control's value to be less than or equal to the provided number.
 *
 * @usageNotes
 *
 * ### Validate against a maximum of 15
 *
 * ```typescript
 * const control = new FormControl(16, max(15));
 *
 * console.log(control.errors); // {max: {max: 15, actual: 16}}
 * ```
 *
 * @returns A validator function that returns an error map with the
 * `max` property if the validation check fails, otherwise `null`.
 *
 */
export function max(max: number, message: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return validate(NgValidators.max(max), 'max', control, message);
  };
}

/**
 * @description
 * Validator that requires the control have a non-empty value.
 *
 * @usageNotes
 *
 * ### Validate that the field is non-empty
 *
 * ```typescript
 * const control = new FormControl('', Validators.required('This field is required'));
 *
 * console.log(control.errors); // {required: {message: 'This field is required'}}
 * ```
 *
 * @returns An error map with the `required` property
 * if the validation check fails, otherwise `null`.
 *
 *
 */

export function required(message: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return validate(NgValidators.required, 'required', control, message);
  };
}

/**
 * @description
 * Validator that requires the control's value be true. This validator is commonly
 * used for required checkboxes.
 *
 * @usageNotes
 *
 * ### Validate that the field value is true
 *
 * ```typescript
 * const control = new FormControl('some value', Validators.requiredTrue('This field is required'));
 *
 * console.log(control.errors); // {required: {message: 'This field is required'}}
 * ```
 *
 * @returns An error map that contains the `required` property
 * set to `true` if the validation check fails, otherwise `null`.
 *
 *
 */

export function requiredTrue(message: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return validate(NgValidators.requiredTrue, 'required', control, message);
  };
}

/**
 * @description
 * Validator that requires the control's value pass an email validation test.
 *
 * Tests the value using a [regular
 * expression](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions)
 * pattern suitable for common use cases. The pattern is based on the definition of a valid email
 * address in the [WHATWG HTML
 * specification](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address) with
 * some enhancements to incorporate more RFC rules (such as rules related to domain names and the
 * lengths of different parts of the address).
 *
 * The differences from the WHATWG version include:
 * - Disallow `local-part` (the part before the `@` symbol) to begin or end with a period (`.`).
 * - Disallow `local-part` to be longer than 64 characters.
 * - Disallow the whole address to be longer than 254 characters.
 *
 * If this pattern does not satisfy your business needs, you can use `Validators.pattern()` to
 * validate the value against a different pattern.
 *
 * @usageNotes
 *
 * ### Validate that the field matches a valid email pattern
 *
 * ```typescript
 * const control = new FormControl('bad@', Validators.email);
 *
 * console.log(control.errors); // {email: {message: 'Invalid email'}}
 * ```
 *
 * @returns An error map with the `email` property
 * if the validation check fails, otherwise `null`.
 *
 */
export function email(message: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return validate(NgValidators.email, 'email', control, message);
  };
}

/**
 * @description
 * Validator that requires the length of the control's value to be greater than or equal
 * to the provided minimum length. This validator is also provided by default if you use the
 * the HTML5 `minlength` attribute. Note that the `minLength` validator is intended to be used
 * only for types that have a numeric `length` property, such as strings or arrays. The
 * `minLength` validator logic is also not invoked for values when their `length` property is 0
 * (for example in case of an empty string or an empty array), to support optional controls. You
 * can use the standard `required` validator if empty values should not be considered valid.
 *
 * @usageNotes
 *
 * ### Validate that the field has a minimum of 3 characters
 *
 * ```typescript
 * const control = new FormControl('ng', Validators.minLength(3));
 *
 * console.log(control.errors); // {minlength: {requiredLength: 3, actualLength: 2}}
 * ```
 *
 * ```html
 * <input minlength="5">
 * ```
 *
 * @returns A validator function that returns an error map with the
 * `minlength` property if the validation check fails, otherwise `null`.
 *
 * @see `updateValueAndValidity()`
 *
 */
export function minLength(minLength: number, message: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return validate(
      NgValidators.minLength(minLength),
      'minlength',
      control,
      message
    );
  };
}

/**
 * @description
 * Validator that requires the length of the control's value to be less than or equal
 * to the provided maximum length. This validator is also provided by default if you use the
 * the HTML5 `maxlength` attribute. Note that the `maxLength` validator is intended to be used
 * only for types that have a numeric `length` property, such as strings or arrays.
 *
 * @usageNotes
 *
 * ### Validate that the field has maximum of 5 characters
 *
 * ```typescript
 * const control = new FormControl('Angular', Validators.maxLength(5));
 *
 * console.log(control.errors); // {maxlength: {requiredLength: 5, actualLength: 7}}
 * ```
 *
 * ```html
 * <input maxlength="5">
 * ```
 *
 * @returns A validator function that returns an error map with the
 * `maxlength` property if the validation check fails, otherwise `null`.
 *
 * @see `updateValueAndValidity()`
 *
 */
export function maxLength(maxLength: number, message: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return validate(
      NgValidators.maxLength(maxLength),
      'maxlength',
      control,
      message
    );
  };
}

/**
 * @description
 * Validator that requires the control's value to match a regex pattern. This validator is also
 * provided by default if you use the HTML5 `pattern` attribute.
 *
 * @usageNotes
 *
 * ### Validate that the field only contains letters or spaces
 *
 * ```typescript
 * const control = new FormControl('1', Validators.pattern('[a-zA-Z ]*'));
 *
 * console.log(control.errors); // {pattern: {requiredPattern: '^[a-zA-Z ]*$', actualValue: '1'}}
 * ```
 *
 * ```html
 * <input pattern="[a-zA-Z ]*">
 * ```
 *
 * ### Pattern matching with the global or sticky flag
 *
 * `RegExp` objects created with the `g` or `y` flags that are passed into `Validators.pattern`
 * can produce different results on the same input when validations are run consecutively. This is
 * due to how the behavior of `RegExp.prototype.test` is
 * specified in [ECMA-262](https://tc39.es/ecma262/#sec-regexpbuiltinexec)
 * (`RegExp` preserves the index of the last match when the global or sticky flag is used).
 * Due to this behavior, it is recommended that when using
 * `Validators.pattern` you **do not** pass in a `RegExp` object with either the global or sticky
 * flag enabled.
 *
 * ```typescript
 * // Not recommended (since the `g` flag is used)
 * const controlOne = new FormControl('1', Validators.pattern(/foo/g));
 *
 * // Good
 * const controlTwo = new FormControl('1', Validators.pattern(/foo/));
 * ```
 *
 * @param pattern A regular expression to be used as is to test the values, or a string.
 * If a string is passed, the `^` character is prepended and the `$` character is
 * appended to the provided string (if not already present), and the resulting regular
 * expression is used to test the values.
 *
 * @returns A validator function that returns an error map with the
 * `pattern` property if the validation check fails, otherwise `null`.
 *
 * @see `updateValueAndValidity()`
 *
 */
export function pattern(pattern: string | RegExp, message: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return validate(NgValidators.pattern(pattern), 'pattern', control, message);
  };
}

export function range(range: [number, number], message?: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    if (isEmptyInputValue(control.value) || isEmptyInputValue(range)) {
      return null; // don't validate empty values to allow optional controls
    }

    const [min, max] = range;
    const value = parseFloat(control.value);

    if (isNaN(value)) {
      return null;
    }

    return value < min || value > max
      ? { range: { range, message, actual: control.value } }
      : null;
  };
}

export function url(message?: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    if (isEmptyInputValue(control.value) || isEmptyInputValue(range)) {
      return null; // don't validate empty values to allow optional controls
    }

    return URL_REGEXP.test(control.value) ? null : { url: { message } };
  };
}

export function equal(value: any, message?: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    if (isEmptyInputValue(control.value) || isEmptyInputValue(value)) {
      return null; // don't validate empty values to allow optional controls
    }

    return value === control.value
      ? null
      : { equal: { requiredValue: value, actual: control.value, message } };
  };
};

export function equalTo(controlNamePath: string, message?: string): ValidatorFn {
  let subscribe: boolean = false;

  return (control: AbstractControl): ValidationErrors | null => {
    const equalControl = control.root.get(controlNamePath);

    if (isEmptyInputValue(control.value) || isEmptyInputValue(controlNamePath) || !equalControl) {
      return null; // don't validate empty values to allow optional controls
    }

    if (!subscribe) {
      subscribe = true;
      equalControl.valueChanges.subscribe(() => {
        control.updateValueAndValidity();
      });
    };

    return equalControl.value === control.value
      ? null
      : { equalTo: { requiredValue: equalControl.value, actual: control.value, message } };
  }
}

export function noWhitespace(message?: string): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    return (control.value || '').trim().length ? null : { whitespace: { message } };
  };
}
